spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -497,37 +494,110 @@ private String resolve(String value) {
 				}
 				else if (exchangeType.equals(ExchangeTypes.TOPIC)) {
 					exchange = new TopicExchange(exchangeName,
 							resolveExpressionAsBoolean(bindingExchange.durable()),
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey, null);
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
+				}
+				else if (exchangeType.equals(ExchangeTypes.HEADERS)) {
+					exchange = new HeadersExchange(exchangeName,
+							resolveExpressionAsBoolean(bindingExchange.durable()),
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
 				}
 				else {
 					throw new BeanInitializationException("Unexpected exchange type: " + exchangeType);
 				}
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, exchange);
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, actualBinding);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+						exchange);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+						actualBinding);
 			}
 		}
 		return queues.toArray(new String[queues.size()]);
 	}
 
+	private Map<String, Object> resolveArguments(Argument[] arguments) {
+		Map<String, Object> map = new HashMap<String, Object>();
+		for (Argument arg : arguments) {
+			String key = resolveExpressionAsString(arg.name(), "name");
+			if (StringUtils.hasText(key)) {
+				Object value = resolveExpression(arg.value());
+				Object type = resolveExpression(arg.type());
+				Class<?> typeClass;
+				String typeName;
+				if (type instanceof Class) {
+					typeClass = (Class<?>) type;
+					typeName = typeClass.getName();
+				}
+				else {
+					Assert.isTrue(type instanceof String, "type must resolve to a Class or String");
+					typeName = (String) type;
+					try {
+						typeClass = ClassUtils.forName(typeName, this.beanClassLoader);
+					}
+					catch (Exception e) {
+						throw new IllegalStateException("Could not load class", e);
+					}
+				}
+				if (value.getClass().getName().equals(typeName)) {
+					if (typeClass.equals(String.class) && !StringUtils.hasText((String) value)) {
+						map.put(key, null);
+					}
+					else {
+						map.put(key, value);
+					}
+				}
+				else {
+					Assert.isTrue(value instanceof String, "value must resolve to a String for type conversion");
+					String valueString = (String) value;
```

Comment:
Unfortunately, we can't do that - we don't know what type of argument value the argument name requires. In this case, yes, TTL is a number but we could match a header, for example `"foo", "10000"` where 10000 is a String.

But, you are right, we don't need the restriction since the CS can handle other conversions. The limit was there when I used SpEL to do the conversion.